From 516efd3b98a51372495696c3ecee526859577267 Mon Sep 17 00:00:00 2001
From: Ben Hutchings <benh@debian.org>
Date: Thu, 30 May 2024 13:14:32 +0100
Subject: [PATCH 289/360] debian: firmware_loader: Log direct loading failures
 as info for d-i

On an installed Debian system, firmware packages will normally be
installed automatically based on a mapping of device IDs to firmware.
Within the Debian installer this has not yet happened and we need a
way to detect missing firmware.

Although many/most drivers log firmware loading failures, they do so
using many different formats.  This adds a single log message to the
firmware loader, which the installer's hw-detect package will look
for.  The log level is set to "info" because some failures are
expected and we do not want to confuse users with bogus error messages
(like in bug #966218).

NOTE: The log message format must not be changed without coordinating
this with the check-missing-firmware.sh in hw-detect.

Link: https://salsa.debian.org/kernel-team/linux/-/blob/debian/6.11/trixie/debian/patches/debian/firmware_loader-log-direct-loading-failures-as-info-for-d-i.path
Signed-off-by: Wentao Guan <guanwentao@uniontech.com>
---
 drivers/base/firmware_loader/main.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/base/firmware_loader/main.c b/drivers/base/firmware_loader/main.c
index c6664a787969..c75b53b13009 100644
--- a/drivers/base/firmware_loader/main.c
+++ b/drivers/base/firmware_loader/main.c
@@ -590,6 +590,10 @@ fw_get_filesystem_firmware(struct device *device, struct fw_priv *fw_priv,
 	}
 	__putname(path);
 
+	if (rc)
+		dev_info(device, "firmware: failed to load %s (%d)\n",
+			 fw_priv->fw_name, rc);
+
 	return rc;
 }
 
-- 
2.20.1

